privacy-protections/storage-blocking/main.js

```diff
@@ -0,0 +1,411 @@
+                    resolve(d.rows[0].value);
+                }, (sql, e) => reject('err - select ' + e.message));
+            });
+
+            return promise;
+        }
+    },
+    {
+        id: 'third party iframe',
+        store: (data) => {
+            let resolve, reject;
+            const promise = new Promise((res, rej) => {resolve = res; reject = rej});
+
+            const iframe = document.createElement('iframe');
+            iframe.src = `${THIRD_PARTY_ORIGIN}/privacy-protections/storage-blocking/iframe.html?data=${data}`;
+            iframe.style.width = '10px';
+            iframe.style.height = '10px';
+            let failTimeout = null;
+
+            function cleanUp(msg) {
+                if (msg.data) {
+                    resolve(msg.data);
+
+                    clearTimeout(failTimeout);
+                    document.body.removeChild(iframe);
+                    window.removeEventListener('message', cleanUp);
+                }
+            }
+
+            window.addEventListener('message', cleanUp);
+            iframe.addEventListener('load', () => {
+                failTimeout = setTimeout(() => reject('timeout'), 1000);
+            });
+
+            document.body.appendChild(iframe);
+
+            return promise;
+        },
+        retrive: () => {
+            let resolve, reject;
+            const promise = new Promise((res, rej) => {resolve = res; reject = rej});
+
+            const iframe = document.createElement('iframe');
+            iframe.src = `${THIRD_PARTY_ORIGIN}/privacy-protections/storage-blocking/iframe.html`;
+            iframe.style.width = '10px';
+            iframe.style.height = '10px';
+            let failTimeout = null;
+
+            function cleanUp(msg) {
+                if (msg.data) {
+                    resolve(msg.data);
+
+                    clearTimeout(failTimeout);
+                    document.body.removeChild(iframe);
+                    window.removeEventListener('message', cleanUp);
+                }
+            }
+
+            window.addEventListener('message', cleanUp);
+            iframe.addEventListener('load', () => {
+                failTimeout = setTimeout(() => reject('timeout'), 1000);
+            });
+
+            document.body.appendChild(iframe);
+
+            return promise;
+        }
+    },
+    {
+        id: 'Cache API',
+        store: (data) => {
+            return caches.open('data').then((cache) => {
+                const res = new Response(data, {
+                    status: 200
+                });
+
+                return cache.put('/cache-api-response', res);
+            });
+        },
+        retrive: () => {
+            return caches.open('data').then((cache) => {
+                return cache.match('/cache-api-response')
+                    .then(r => r.text());
+            });
+        }
+    },
+    {
+        id: 'browser cache',
+        store: (data) => {
+            // already done before all tests
```

Comment:
see `storeData()`. We can't really pass a number to server (e.g. `/cached-random-number?number=${data}`) because we won't be able to guess that url later on. We have to have one url (`/cached-random-number`) that we will query to warm up the cache and retrieve the value. And for that we have to do this operation first and use the number server produces to fill all other storage.